lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -1,94 +1,482 @@
+        ParamType Enum initializer: this uses a mixed-case name string in
+        addition to the conversion method simply because with only the
+        Callable value I ran into naming issues.
+
+        TODO: I'll experiment a bit more with other ways to avoid the
+        problems I encountered, as I'm not actually using this. Perhaps
+        using auto() for the first parameter would work?
+        """
+        self.friendly = name
+        self.converter = converter
+
+    def __str__(self) -> str:
+        return self.name
+
+    def convert(self, value: Any) -> Any:
+        """
+        convert Convert a string value to the expected type
+
+        Args:
+            value: The value of an input parameter
+
+        Returns:
+            Appropriate translated value
+
+        Raises:
+            Any conversion errors
+        """
+        return self.converter(value)
+
+
+class Parameter:
+    """
+    Define the attributes of a parameter using the ParamType ENUM
+
+    Note that a parameter that's "required" must also be non-empty.
+    """
+
+    def __init__(self, name: AnyStr, type: ParamType, required: bool = False):
+        """
+        __init__ Initialize a Parameter object describing a JSON parameter
+        with its type and attributes.
+
+        Args:
+            name: Parameter name
+            type: Parameter type
+            required: whether the parameter is required (default to False)
+        """
+        self.name = name
+        self.type = type
+        self.required = required
+
+    def __str__(self) -> str:
+        return f"Parameter<{self.name}:{self.type}>"
+
+
+class Schema:
+    """
+    Define the client input schema for a server query that's based on
+    Elasticsearch.
+    """
+
+    def __init__(self, *parameters: Parameter):
+        """
+        __init__ Specify an interface schema as a list of Parameter objects.
+
+        Args:
+            parameters: a list of Parameter objects
+        """
+        self.parameters = {p.name: p for p in parameters}
+
+    def get(self, name: AnyStr) -> Parameter:
+        """
+        get Retrieve a parameter by name
+
+        Args:
+            name: Parameter name to return
+
+        Raises:
+            KeyError: The parameter wasn't found
+
+        Returns:
+            the specified Parameter value
+        """
+        if name not in self.parameters:
+            raise KeyError(f"Parameter {name} not found")
+        return self.parameters[name]
+
+    def __iter__(self) -> Iterable[Parameter]:
+        """
+        __iter__ Iterating on a Schema iterates through the list of parameters
+
+        Returns:
+            Parameter list iterator
+        """
+        return iter(self.parameters.items())
+
+    def __str__(self) -> str:
+        return f"Schema<{self.parameters}>"
+
+
+class ElasticBase(Resource):
+    """
+    ElasticBase A base class for Elasticsearch queries that allows subclasses
+    to provide customers pre- and post- processing.
```

Comment:
Except that I had to scratch my head as to why this is a Resource and not just a regular object, but maybe that is just me.

Reply:
A `flask_restful` `Resource` object does something, but I can't tell what this class is adding to the `Resource` class very easily.

Reply:
It is a subclass of `Resource` which provides `post()` and `get()` methods so that it can implement an endpoint in a RESTful interface (or, I guess, so that it can be a base class which can be further extended to implement such endpoints...).